presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -644,6 +649,100 @@ private void validateMaterialziedViewQueryPlan(Statement query)
             validator.process(query, new MaterializedViewPlanValidatorContext());
         }
 
+        @Override
+        protected Scope visitRefreshMaterializedView(RefreshMaterializedView node, Optional<Scope> scope)
+        {
+            analysis.setUpdateType("INSERT");
+
+            QualifiedObjectName viewName = createQualifiedObjectName(session, node.getTarget(), node.getTarget().getName());
+
+            ConnectorMaterializedViewDefinition view = metadata.getMaterializedView(session, viewName)
+                    .orElseThrow(() -> new SemanticException(MISSING_MATERIALIZED_VIEW, node, "Materialized view '%s' does not exist", viewName));
+
+            accessControl.checkCanInsertIntoTable(session.getRequiredTransactionId(), session.getIdentity(), session.getAccessControlContext(), viewName);
+
+            // Analyzer will check for SELECT permission on materialized view table, but REFERSH MATERIALIZED VIEW only needs INSERT permission.
+            // So use a new StatementAnalyzer instance with access check disabled to analyze the materialized view table.
+            StatementAnalyzer analyzer = new StatementAnalyzer(
+                    analysis,
+                    metadata,
+                    sqlParser,
+                    new AllowAllAccessControl(),
+                    session,
+                    warningCollector);
+            Scope viewScope = analyzer.analyze(node.getTarget(), scope);
+
+            Map<SchemaTableName, Expression> tablePredicates = extractTablePredicates(viewName, node.getWhere(), viewScope, metadata, session);
```

Comment:
may be I am missing something, I tried to run this for a simple test case [https://github.com/prestodb/presto/blob/5706c39f5697d7d1a4dacb55e439ec6f863bfc04/presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java#L1043](url)

With insert replaced with refresh as following:
`            queryRunner.execute("REFRESH MATERIALIZED VIEW test_orders_view where ds='2020-01-01'");
`

It does not work, can you please check. I will try to debug bit more later.

Reply:
I debugged it bit more. The main issue is here:
`
 Query refreshQuery = tablePredicates.containsKey(toSchemaTableName(viewName)) ?                    buildQueryWithPredicate(viewQuery, tablePredicates.get(toSchemaTableName(viewName)))                    : viewQuery;<br class="Apple-interchange-newline">`

We generating predicates against the base tables, which is correct. 
We need to generate the predicate stitched query against the base tables, not against the view. 
We can quickly and easily leverage PredicateStitcher class to fix the original sql query to stitch the newly generated predicates.

Reply:
Base table predicate stitching is done in analyzeBaseTableForRefreshMaterializedView(). Here is to handle the view partitions specified in the where clause that don't have a direct mapping to any base table partition and thus don't belong to any base table. We attach such partition predicate to the view query.